deps/rabbit/src/rabbit_message_interceptor.erl

```diff
@@ -1,50 +1,43 @@
-%% This Source Code Form is subject to the terms of the Mozilla Public
-%% License, v. 2.0. If a copy of the MPL was not distributed with this
-%% file, You can obtain one at https://mozilla.org/MPL/2.0/.
-%%
-%% Copyright (c) 2007-2025 Broadcom. All Rights Reserved. The term “Broadcom” refers to Broadcom Inc. and/or its subsidiaries. All rights reserved.
-
-%% This module exists since 3.12 replacing plugins rabbitmq-message-timestamp
-%% and rabbitmq-routing-node-stamp. Instead of using these plugins, RabbitMQ core can
-%% now be configured to add such headers. This enables non-AMQP 0.9.1 protocols (that
-%% do not use rabbit_channel) to also add AMQP 0.9.1 headers to incoming messages.
 -module(rabbit_message_interceptor).
--include("mc.hrl").
-
--export([intercept/1]).
 
--define(HEADER_TIMESTAMP, <<"timestamp_in_ms">>).
--define(HEADER_ROUTING_NODE, <<"x-routed-by">>).
+-export([
+    intercept/3,
+    set_msg_annotation/4
+]).
 
--spec intercept(mc:state()) -> mc:state().
-intercept(Msg) ->
-    Interceptors = persistent_term:get(incoming_message_interceptors, []),
-    lists:foldl(fun({InterceptorName, Overwrite}, M) ->
-                        intercept(M, InterceptorName, Overwrite)
-                end, Msg, Interceptors).
+-callback intercept(
+    Msg :: mc:state(),
+    MsgInterceptorCtx :: map(),
+    Config :: #{atom() := term()}
+) -> mc:state().
 
-intercept(Msg, set_header_routing_node, Overwrite) ->
-    Node = atom_to_binary(node()),
-    set_annotation(Msg, ?HEADER_ROUTING_NODE, Node, Overwrite);
-intercept(Msg0, set_header_timestamp, Overwrite) ->
-    Ts = mc:get_annotation(?ANN_RECEIVED_AT_TIMESTAMP, Msg0),
-    Msg = set_annotation(Msg0, ?HEADER_TIMESTAMP, Ts, Overwrite),
-    set_timestamp(Msg, Ts, Overwrite).
+-spec intercept(Msg, MsgInterceptorCtx, MsgDirection) -> Resp when
+    Msg :: mc:state(),
+    MsgInterceptorCtx :: map(),
+    MsgDirection:: incoming | outgoing,
+    Resp :: mc:state().
+intercept(Msg, MsgInterceptorCtx, MsgDirection) ->
+    InterceptorsList = list_to_atom(atom_to_list(MsgDirection)
+                                    ++ "_message_interceptors"),
+    Interceptors = persistent_term:get(InterceptorsList, []),
+    lists:foldl(fun({Module, Config}, Msg0) ->
+        try
+            Module:intercept(Msg0, MsgInterceptorCtx, Config)
```

Comment:
I feel the direction should be passed to the interceptor, you may want to just have a single interceptor for both incoming and outgoing.

Reply:
I assumed an interceptor would do the same independently of the direction of the message. It is the user the one that decides (if the interceptor allows it by providing the appropriate `message_interceptors.<direction>.` cuttlefish mapping) in which direction should the interceptor be applied, or if it should be applied in both.

Reply:
As suggested by @gomoripeti , the `rabbit_message_interceptor` module could also add the message direction in the `MsgInterceptorCtx` map. How does that sound?

Reply:
that sounds good.